crates/hulk_replayer/src/window.rs

```diff
@@ -92,6 +106,59 @@ fn join_timing(replayer: &Replayer<ReplayerHardwareInterface>) -> FrameRange {
                 .map(|timing| timing.timestamp + timing.duration)
         })
         .max()
-        .expect("there isn't any index that contains at least one frame");
+        .unwrap_or(SystemTime::UNIX_EPOCH);
     FrameRange::new(AbsoluteTime::new(begin), AbsoluteTime::new(end))
 }
+
+fn spawn_replay_thread(
+    replayer: Arc<Mutex<Replayer<ReplayerHardwareInterface>>>,
+    egui_context: Context,
+    mut time: watch::Receiver<SystemTime>,
+    frames_changed: watch::Sender<()>,
+) {
+    spawn(move || {
+        loop {
+            let mut frames_found = false;
+            let mut replayer = replayer.lock().unwrap();
+            let mut indices = replayer.get_recording_indices_mut();
+            for index in indices.values_mut() {
+                match index.collect_next_frame_metadata() {
+                    Ok(Some(())) => {
+                        frames_changed.send(()).unwrap();
+                        frames_found = true;
+                    }
+                    Ok(None) => {}
+                    Err(error) => eprintln!("{error}"),
+                }
+            }
+            if !frames_found {
+                break;
+            }
+            if time.has_changed().unwrap() {
+                time.mark_unchanged();
```

Comment:
This looks like a lost wake-up? After checking it could change again but afterwards you discard it by marking it as unchanged.

Reply:
This is a watch channel, so we only care about the last value.
So lost updates are okay as long as we get the last value before a long break.
Note that I check the flag, remove the flag, then inspect the value.
If there is another update between clearing the flag and reading the value, worst case is simulating the same value twice, since the flag is then set again in the next iteration.

Reply:
My point was an update between checking and removing the flag. After checking, you have decided that you want to clear the flag regardless if there was another update.

Reply:
That's a lost update, not a lost wakeup, since in that case too we use the newer value.